joomla/joomla-cms
# [3.9] Pre-Update Checker

### Summary of Changes

Based on https://github.com/joomla/joomla-cms/pull/16494 we continued developing the Pre-Update Checker for Joomla! 3.9. Like the Joomla! installation process, the com_joomlaupdate component now checks PHP options, settings and extension compatibility.
The evaluation process is currently a copy of the installation process pre-checker.
The translation of extension names is duplicated from com_installer.
Future PRs should address these issues by moving these functions to a core library.

#### Conditional checks:
##### If next J! version >= 4:
* Check for unsupported database types (sqlsrv, sqlazure)
##### If PHP version of next J! < 7:
* Magic Quotes (+ Runtime)
* Register Globals
* Safe Mode

### Testing Instructions

These instructions are based on a Joomla 3.9 installation.
Change values accordingly.
The path <code>/</code> represents the Joomla! installation directory.

#### Test Joomla! Pre-Update Check

1. Create file `list.xml` at `/`
```xml
<extensionset>
	<extension element="joomla" type="file" version="4.0.0" targetplatformversion="3.9" detailsurl="http://localhost/extension_sts.xml" />
</extensionset>
```

2. Create file `extension_sts.xml` at `/`
```xml
<updates>
	<update>
		<name>Joomla! 3.7</name>
		<description>Joomla! 3.7 CMS</description>
		<element>joomla</element>
		<type>file</type>
		<version>4.0.0</version>
		<infourl title="Joomla!">https://www.joomla.org/announcements/release-news/5711-joomla-3-7-5-release.html</infourl>
		<downloads>
			<downloadurl type="full" format="zip">https://downloads.joomla.org/cms/joomla3/3-7-5/Joomla_3.7.5-Stable-Update_Package.zip</downloadurl>
		</downloads>
		<tags>
			<tag>stable</tag>
		</tags>
		<maintainer>Joomla! PLT</maintainer>
		<maintainerurl>https://www.joomla.org</maintainerurl>
		<section>STS</section>
		<targetplatform name="joomla" version="3.[23456789]" />
		<php_minimum>7.0</php_minimum>
	</update>
</updates>
```

3. Components -> Joomla! Update -> Options

Custom URL: `localhost/list.xml`

4. Save & Close

There should be a fake update to Joomla 4.

5. Change your PHP or webserver settings to trigger a failed compatibility check.

`display_errors = On`

`file_uploads = Off`

`output_buffering = On`

6. Run `chmod -w configuration.php` to disable write on the configuration file.

7. Visit the update site and check the compatibilty results.

#### Test Extension Compatibility

1. Install some extensions, which are not compatible with your hacked Joomla! target version (4.0.0).

2. Change the XML Update-URL of any incompatible extension in the table `#__update_sites` in the column `location` to an own custom XML file.

3. Put the following contents in your XML update file. Replace version 1.5 with the actual installed version of your selected extension.

```xml
<updates>
	<update>
		<version>1.5</version>
		<downloads>
			<downloadurl type="full" format="zip">https://fakeurl</downloadurl>
		</downloads>
		<targetplatform name="joomla" version="4.0.0" />
	</update>
</updates>
```

4. Check the extensions compatiblity result.

5. Change the column `manifest_cache` in the table `#__extensions` of the hacked extension by changing the field `version` to an older one. This should trigger a compatiblilty, but based on a newer version.

#### Test Database compatibility

Add your current database type (e.g. `mysqli`) to `$unsupportedDatabaseTypes` in function `isDatabaseTypeSupported()` in file `./administrator/components/com_joomlaupdate/models/default.php`.

### Expected result

You should see a full report on compatibility of the update. PHP options, settings and extensions should be marked accordingly.
An incompatible database type should cause the message "Your host doesn't support the minimum requirements for Joomla".
<img width="1037" alt="preupdate1" src="https://user-images.githubusercontent.com/11752441/41082853-eb9eb212-6a2f-11e8-99dc-82840f6f4429.png">
<img width="830" alt="preupdate2" src="https://user-images.githubusercontent.com/11752441/41082858-ed6607c6-6a2f-11e8-965e-a765cda8bd2a.png">

### Actual result

No results, since this is a new feature.

### Documentation Changes Required

The following page should be updated with recent screenshots and descriptions.
https://docs.joomla.org/J3.x:Updating_from_an_existing_version

Developed @icampus, @SamuelSchepp, @Kev1n337, @DrDreave
